srv/salt/_modules/cephprocesses.py

```diff
@@ -49,101 +50,274 @@
-                        if pdict_exe in res['up']:
-                            res['up'][pdict_exe] = res['up'][pdict_exe] + [pdict_pid]
-                        else:
-                            res['up'][pdict_exe] = [pdict_pid]
-
-            if role in absent_processes.keys():
-                for proc in absent_processes[role]:
-                    if proc in res['up']:
-                        # running is deceptive here
-                        # running indicates wheter the service is in it's expected state
-                        running = False
-                        # pylint: disable=line-too-long
-                        log.error("ERROR: process {} for role {} is pending(working)".format(proc, role))
-            else:
-                for proc in processes[role]:
-                    if proc not in res['up']:
-                        if not quiet:
-                            # pylint: disable=line-too-long
-                            log.error("ERROR: process {} for role {} is not running".format(proc, role))
-                        running = False
-                        res['down'] += [proc]
-
-            # pylint: disable=fixme
-            # FIXME: Map osd.ids to processes.pid to improve qualitify of logging
-            # currently you can only say how many osds/ if any are down, but not
-            # which osd is down exactly.
-            if role == 'storage':
-                if 'ceph-osd' in res['up']:
-                    if len(__salt__['osd.list']()) > len(res['up']['ceph-osd']):
-                        if not quiet:
-                            log.error("ERROR: At least one OSD is not running")
-                        res = {'up': {}, 'down': {'ceph-osd': 'ceph-osd'}}
-                        running = False
-
-    return res if results else running
-
-
-# pylint: disable=unused-argument
-def down(**kwargs):
+
+def check(results=False, **kwargs):
+
+    """
+    Query the status of running processes for each role.  Return False if any
+    fail.  If results flag is set, return a dictionary of the form:
+      { 'down': [ process, ... ], 'up': { process: [ pid, ... ], ...} }
+    """
+    _extend_processes()
+    res = MetaCheck(**kwargs)
+
+    if 'roles' not in __pillar__:
+        log.error("Did not find _roles_ in pillar. Aborting")
+        return False
+    for role in kwargs.get('roles', __pillar__['roles']):
+        for running_proc in psutil.process_iter():
+            res.add(ProcInfo(running_proc), role)
+        res.check_inverts(role)
+        res.check_absents(role)
+        if role == 'storage':
+            res.check_osds()
+    return res.report() if results else res.running
+
+
+def down():
     """
     Based on check(), return True/False if all Ceph processes that are meant
     to be running on a node are down.
     """
-    return True if not list(check(True, True)['up'].values()) else False
+    return True if not list(check(True)['up'].values()) else False
```

Comment:
Nitpick: I think we could simplify this like this: `return not list(check(True)['up'].values())`, but it is fine either way (and your version is more explicit, which does not hurt)

Reply:
right, this is readability vs brevity. Since we are not line-wrapping here, I personally prefer the former over the latter.